lib/appium.js

```diff
@@ -246,10 +248,16 @@ class AppiumDriver extends BaseDriver {
    * @return {Array} Unique session ID and capabilities
    */
   async createSession (jsonwpCaps, reqCaps, w3cCapabilities) {
-    const {defaultCapabilities} = this.args;
+    let {defaultCapabilities} = this.args;
+    defaultCapabilities = _.cloneDeep(defaultCapabilities);
+    const defaultSettings = pullSettings(defaultCapabilities);
+    jsonwpCaps = _.cloneDeep(jsonwpCaps);
+    const jwpSettings = Object.assign({}, defaultSettings, pullSettings(jsonwpCaps));
+    w3cCapabilities = _.cloneDeep(w3cCapabilities);
+    const w3cSettings = Object.assign({}, jwpSettings, pullSettings(w3cCapabilities));
```

Comment:
why are we merging jwp settings into w3c settings? below these seem like exclusive options

Reply:
I've seen the cases where the client sends jsonwp caps only, but the request contains firstMatch/alwaysMatch props. In such case we assume the client supports W3C and merge jwp caps to w3c ones